querybook/webapp/components/DataDocGitHub/GitHubModal.tsx

```diff
@@ -0,0 +1,82 @@
+import React, { useCallback, useState } from 'react';
+
+import { ComponentType, ElementType } from 'const/analytics';
+import { trackClick } from 'lib/analytics';
+import { GitHubResource, IGitHubAuthResponse } from 'resource/github';
+import { Message } from 'ui/Message/Message';
+import { Modal } from 'ui/Modal/Modal';
+
+import { GitHubAuth } from './GitHubAuth';
+
+interface IProps {
+    docId: number;
+    isAuthenticated: boolean;
+    setIsAuthenticated: (isAuthenticated: boolean) => void;
+    onClose: () => void;
+}
+
+export const GitHubModal: React.FunctionComponent<IProps> = ({
+    docId,
+    isAuthenticated,
+    setIsAuthenticated,
+    onClose,
+}) => {
+    const [errorMessage, setErrorMessage] = useState<string>(null);
+
+    const handleConnectGitHub = useCallback(async () => {
+        trackClick({
+            component: ComponentType.DATADOC_PAGE,
+            element: ElementType.GITHUB_CONNECT_BUTTON,
+        });
+
+        try {
+            const { data }: { data: IGitHubAuthResponse } =
+                await GitHubResource.connectGithub();
+            const url = data.url;
+            if (!url) {
+                throw new Error('Failed to get GitHub authentication URL');
+            }
+            const authWindow = window.open(url);
+
+            const receiveMessage = () => {
+                authWindow.close();
+                delete window.receiveChildMessage;
+                window.removeEventListener('message', receiveMessage, false);
+                setIsAuthenticated(true);
+            };
+            window.receiveChildMessage = receiveMessage;
+
+            // If the user closes the authentication window manually, clean up
+            const timer = setInterval(() => {
```

Comment:
why it needs a timer here?

Reply:
detects if the user manually closed the authentication window to throw an error authentication process incomplete